File: apps/mobile/modules/safely-masked-input/android/src/main/java/expo/modules/safelymaskedinput/SafelyMaskedInputModule.kt
package expo.modules.safelymaskedinput

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class ValueUpdate : Record {
    @Field
    var text: String? = null

    @Field
    var eventCount: Int = 0
}

class SafelyMaskedInputModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("SafelyMaskedInput")

        Function("focus") { viewTag: Int ->
            findView(viewTag)?.focusInput()
        }

        Function("blur") { viewTag: Int ->
            findView(viewTag)?.blurInput()
        }

        Function("setText") { viewTag: Int, text: String ->
            findView(viewTag)?.setRawValue(text)
        }

        Function("setCursorPosition") { viewTag: Int, position: Int ->
            findView(viewTag)?.setCursorPos(position)
        }

        View(SafelyMaskedInputView::class) {
            Events("onChangeText", "onFocusChange", "onPaste")

            Prop("decimals") { view: SafelyMaskedInputView, decimals: Int ->
                view.setDecimals(decimals)
            }

            Prop("decimalSeparator") { view: SafelyMaskedInputView, separator: String ->
                view.setDecimalSeparator(separator)
            }

            Prop("value") { view: SafelyMaskedInputView, update: ValueUpdate? ->
                update?.let { view.setValueUpdate(it) }
            }

            Prop("fontSize") { view: SafelyMaskedInputView, size: Double ->
                view.setFontSizeValue(size.toFloat())
            }

            Prop("fontFamily") { view: SafelyMaskedInputView, family: String ->
                view.setFontFamilyValue(family)
            }

            Prop("textColor") { view: SafelyMaskedInputView, color: String ->
                view.setTextColorValue(color)
            }

            Prop("placeholder") { view: SafelyMaskedInputView, placeholder: String ->
                view.setPlaceholderValue(placeholder)
            }

            Prop("placeholderTextColor") { view: SafelyMaskedInputView, color: String ->
                view.setPlaceholderTextColorValue(color)
            }

            Prop("keyboardType") { view: SafelyMaskedInputView, type: String ->
                view.setKeyboardTypeValue(type)
            }

            Prop("editable") { view: SafelyMaskedInputView, editable: Boolean ->
                view.setEditableValue(editable)
            }

            Prop("autoFocus") { view: SafelyMaskedInputView, autoFocus: Boolean ->
                view.setAutoFocusValue(autoFocus)
            }

            Prop("integerColor") { view: SafelyMaskedInputView, color: String ->
                view.setIntegerColorValue(color)
            }

            Prop("integerOpacity") { view: SafelyMaskedInputView, opacity: Double ->
                view.setIntegerOpacityValue(opacity)
            }

            Prop("decimalColor") { view: SafelyMaskedInputView, color: String ->
                view.setDecimalColorValue(color)
            }

            Prop("decimalOpacity") { view: SafelyMaskedInputView, opacity: Double ->
                view.setDecimalOpacityValue(opacity)
            }

            Prop("placeholderDigitColor") { view: SafelyMaskedInputView, color: String ->
                view.setPlaceholderDigitColorValue(color)
            }

            Prop("placeholderDigitOpacity") { view: SafelyMaskedInputView, opacity: Double ->
                view.setPlaceholderDigitOpacityValue(opacity)
            }

            Prop("suffix") { view: SafelyMaskedInputView, value: String ->
                view.setSuffixValue(value)
            }

            Prop("suffixColor") { view: SafelyMaskedInputView, color: String ->
                view.setSuffixColorValue(color)
            }

            Prop("suffixOpacity") { view: SafelyMaskedInputView, opacity: Double ->
                view.setSuffixOpacityValue(opacity)
            }

            Prop("suffixFontSize") { view: SafelyMaskedInputView, size: Double ->
                view.setSuffixFontSizeValue(size.toFloat())
            }
        }
    }

    private fun findView(viewTag: Int): SafelyMaskedInputView? {
        return appContext.findView<SafelyMaskedInputView>(viewTag)
    }
}

File: apps/mobile/modules/safely-masked-input/ios/SafelyMaskedInputModule.swift
import ExpoModulesCore

struct ValueUpdate: Record {
    @Field var text: String? = nil
    @Field var eventCount: Int = 0
}

public class SafelyMaskedInputModule: Module {
    public func definition() -> ModuleDefinition {
        Name("SafelyMaskedInput")

        Function("focus") { (viewTag: Int) in
            DispatchQueue.main.async { [weak self] in
                guard let view = self?.findView(viewTag) else { return }
                view.focusInput()
            }
        }

        Function("blur") { (viewTag: Int) in
            DispatchQueue.main.async { [weak self] in
                guard let view = self?.findView(viewTag) else { return }
                view.blurInput()
            }
        }

        Function("setText") { (viewTag: Int, text: String) in
            DispatchQueue.main.async { [weak self] in
                guard let view = self?.findView(viewTag) else { return }
                view.setRawValue(text)
            }
        }

        Function("setCursorPosition") { (viewTag: Int, position: Int) in
            DispatchQueue.main.async { [weak self] in
                guard let view = self?.findView(viewTag) else { return }
                view.setCursorPos(position)
            }
        }

        View(SafelyMaskedInputView.self) {
            Events("onChangeText", "onFocusChange", "onPaste")

            Prop("decimals") { (view: SafelyMaskedInputView, decimals: Int) in
                view.setDecimals(decimals)
            }

            Prop("decimalSeparator") { (view: SafelyMaskedInputView, separator: String) in
                view.setDecimalSeparator(separator)
            }

            Prop("value") { (view: SafelyMaskedInputView, update: ValueUpdate?) in
                guard let update else { return }
                view.setValueUpdate(update)
            }

            Prop("fontSize") { (view: SafelyMaskedInputView, size: Double) in
                view.setFontSizeValue(CGFloat(size))
            }

            Prop("fontFamily") { (view: SafelyMaskedInputView, family: String) in
                view.setFontFamilyValue(family)
            }

            Prop("textColor") { (view: SafelyMaskedInputView, color: String) in
                view.setTextColorValue(color)
            }

            Prop("placeholder") { (view: SafelyMaskedInputView, placeholder: String) in
                view.setPlaceholderValue(placeholder)
            }

            Prop("placeholderTextColor") { (view: SafelyMaskedInputView, color: String) in
                view.setPlaceholderTextColorValue(color)
            }

            Prop("keyboardType") { (view: SafelyMaskedInputView, type: String) in
                view.setKeyboardTypeValue(type)
            }

            Prop("editable") { (view: SafelyMaskedInputView, editable: Bool) in
                view.setEditableValue(editable)
            }

            Prop("autoFocus") { (view: SafelyMaskedInputView, autoFocus: Bool) in
                view.setAutoFocusValue(autoFocus)
            }

            Prop("integerColor") { (view: SafelyMaskedInputView, color: String) in
                view.setIntegerColorValue(color)
            }

            Prop("integerOpacity") { (view: SafelyMaskedInputView, opacity: Double) in
                view.setIntegerOpacityValue(opacity)
            }

            Prop("decimalColor") { (view: SafelyMaskedInputView, color: String) in
                view.setDecimalColorValue(color)
            }

            Prop("decimalOpacity") { (view: SafelyMaskedInputView, opacity: Double) in
                view.setDecimalOpacityValue(opacity)
            }

            Prop("placeholderDigitColor") { (view: SafelyMaskedInputView, color: String) in
                view.setPlaceholderDigitColorValue(color)
            }

            Prop("placeholderDigitOpacity") { (view: SafelyMaskedInputView, opacity: Double) in
                view.setPlaceholderDigitOpacityValue(opacity)
            }

            Prop("suffix") { (view: SafelyMaskedInputView, value: String) in
                view.setSuffixValue(value)
            }

            Prop("suffixColor") { (view: SafelyMaskedInputView, color: String) in
                view.setSuffixColorValue(color)
            }

            Prop("suffixOpacity") { (view: SafelyMaskedInputView, opacity: Double) in
                view.setSuffixOpacityValue(opacity)
            }

            Prop("suffixFontSize") { (view: SafelyMaskedInputView, size: Double) in
                view.setSuffixFontSizeValue(CGFloat(size))
            }
        }
    }

    private func findView(_ viewTag: Int) -> SafelyMaskedInputView? {
        guard let appContext = appContext else { return nil }
        return appContext.findView(withTag: viewTag, ofType: SafelyMaskedInputView.self)
    }
}
